aspects/schema_test.go

```diff
@@ -1680,3 +1682,154 @@ func (*schemaSuite) TestArrayFailsWithBadUniqueType(c *C) {
 	_, err := aspects.ParseSchema(schemaStr)
 	c.Assert(err, ErrorMatches, `cannot parse array's "unique" constraint: json: cannot unmarshal string into Go value of type bool`)
 }
+
+func (*schemaSuite) TestErrorContainsPathPrefixes(c *C) {
+	schemaStr := []byte(`{
+	"schema": {
+		"foo": {
+			"schema": {
+				"bar": {
+					"schema": {
+						"baz": "string"
+					}
+				}
+			}
+		}
+	}
+}`)
+
+	schema, err := aspects.ParseSchema(schemaStr)
+	c.Assert(err, IsNil)
+
+	type testcase struct {
+		name  string
+		input []byte
+		err   string
+	}
+
+	testcases := []testcase{
+		{
+			name:  "top level",
+			input: []byte(`{"bar": 1}`),
+			err:   `cannot validate top level element: map contains unexpected key "bar"`,
+		},
+		{
+			name:  "1 level of nesting",
+			input: []byte(`{"foo": {"baz": 1}}`),
+			err:   `cannot validate element under "foo": map contains unexpected key "baz"`,
+		},
+		{
+			name:  "2 levels of nesting",
+			input: []byte(`{"foo": {"bar": {"boo": 1}}}`),
+			err:   `cannot validate element under "foo.bar": map contains unexpected key "boo"`,
+		},
+	}
+
+	for _, tc := range testcases {
+		err = schema.Validate(tc.input)
+		c.Assert(err, ErrorMatches, tc.err, Commentf("test case %q failed", tc.name))
+	}
+}
+
+func (*schemaSuite) TestPathPrefixWithMapUnderUserType(c *C) {
+	schemaStr := []byte(`{
+	"types": {
+		"my-type": {
+			"schema": {
+				"bar": {
+					"type": "int",
+					"min": 0
+				}
+			}
+		}
+	},
+	"schema": {
+		"foo": "$my-type"
+	}
+}`)
+
+	schema, err := aspects.ParseSchema(schemaStr)
+	c.Assert(err, IsNil)
+
+	input := []byte(`{"foo": {"bar": -1}}`)
+	err = schema.Validate(input)
+	c.Assert(err, ErrorMatches, `cannot validate element under "foo.bar": -1 is less than the allowed minimum 0`)
+}
+
+func (*schemaSuite) TestPathPrefixWithArrayUnderUserType(c *C) {
+	schemaStr := []byte(`{
+	"types": {
+		"my-type": {
+			"type": "int",
+			"min": 0
+		}
+	},
+	"schema": {
+		"foo": {
+			"type": "array",
+			"values": "$my-type"
+		}
+	}
+}`)
+	schema, err := aspects.ParseSchema(schemaStr)
+	c.Assert(err, IsNil)
+
+	input := []byte(`{"foo": [-1]}`)
+	err = schema.Validate(input)
+	c.Assert(err, ErrorMatches, `cannot validate element under "foo": -1 is less than the allowed minimum 0`)
+}
+
+func (*schemaSuite) TestPathPrefixWithArrayUnderUserWithAContainerElementType(c *C) {
+	schemaStr := []byte(`{
+	"types": {
+		"my-type": {
+			"type": "array",
+			"values": {
+				"schema": {
+					"bar": {
+						"type": "int",
+						"min": 0
+					}
+				}
+			}
+		}
+	},
+	"schema": {
+		"foo": "$my-type"
+	}
+}`)
+	schema, err := aspects.ParseSchema(schemaStr)
+	c.Assert(err, IsNil)
+
+	input := []byte(`{"foo": [{"bar": -1}]}`)
+	err = schema.Validate(input)
+	c.Assert(err, ErrorMatches, `cannot validate element under "foo.bar": -1 is less than the allowed minimum 0`)
```

Comment:
Wondering a little bit about this schema path notation, and ease of pinpointing errors for more complex scenarios, OTOH, pretty good vs complexity.

Reply:
Yeah, I had the same feeling and was looking for feedback on this exact thing :) it's fine for maps but it might not be very helpful if there are many nested or large arrays. One approach I thought we could take was to store the path with some type of placeholder for arrays that would then be filled in when building the validation error. It does mean we have to pass around more data but it's feasible. So the output would look like `cannot accept element under "foo.[0].bar: ...` which seems nicer